components/client/src/main/java/com/hotels/styx/client/StyxHttpClient.java

```diff
@@ -44,39 +47,96 @@
     private static final int DEFAULT_HTTPS_PORT = 443;
     private static final int DEFAULT_HTTP_PORT = 80;
 
-    private final Optional<String> userAgent;
-    private final ConnectionSettings connectionSettings;
-    private final int maxResponseSize;
-    private final Connection.Factory connectionFactory;
-    private final boolean isHttps;
-
-    private StyxHttpClient(Builder builder, boolean isHttps) {
-        this.userAgent = Optional.ofNullable(builder.userAgent);
-        this.connectionSettings = builder.connectionSettings;
-        this.maxResponseSize = builder.maxResponseSize;
-        this.connectionFactory = builder.connectionFactory;
-        this.isHttps = isHttps;
+    private final Builder transactionParameters;
```

Comment:
Storing the builder violates immutability.

Reply:
It does. But it saves lots of repeated code.

The `transactionParameters` stores the default client settings at the HTTP client. These settings are inherited into HTTP request transaction where the parameters can be overridden. Storing them as a builder offers a convenient way of inheriting the default settings into a transaction. For two reasons:

* avoids copying parameters into auxiliary data structures
* any copying related to parameter inheritance can be encapsulated in the builder class itself

Perhaps I'll need to rename `transactionParameters` to `defaultClientConfiguration`.

Reply:
We have been consistent in our policy toward immutability so far, we cannot justify violating that just to type less code.

Besides, preserving immutability is the entire purpose of using the builder pattern. If we wanted to write mutable code we wouldn't use builders at all.

As a way to reduce code without introducing mutability, we could `clone` the builder, though there would still be internal mutability.

Reply:
The builder pattern in fact in Java is used as a substitute for optional constructor arguments (See Effective Java by Joshua Bloch). The problem with Java is that the constructors don't scale well with large numbers of parameters. It has little to do with immutability (although some).

Having said that. The class at the moment indeed has a problem with immutability. But this is not a problem with storing the builder itself. The problem is that the builder is not cloned (as you perhaps noticed?) before it is passed in to the `StyxHttpClient` constructor.

I will add a call to the `clone` method.

Reply:
Why don't you use the Builder(builder) constructor that the Builder provides instead of clone()?  Cloneable is not something we want to encourage and I would say it's been strongly advised against for years... 
Other than that, Builder is used to avoid creating 'incomplete objects'  (that's why we cannot use 'Java Beans' for that) and also avoid the limitations of constructors and static factories (that they do not provide either named or optional parameters) so we can construct objects in a clearer/more friendly way (it's usually complex objects, but we are using it in most cases). Even though it's true that it's one of the advantages of Builders, immutability can be achieved by using constructors or static factories...

Reply:
Hey all. I mistakenly referred to `clone`. The `Builder` already implements a method called `copy` that replicates the builder object. The client is instantiated from the `build` object by:

```
            return new StyxHttpClient(connectionFactory, this.copy());
```

Apologies for the confusion.

Reply:
The client class is now immutable from the outside. At least in terms of `Builder` object.